File: clients/go/examples/collections-find-many-iterate-manual.go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/datastax/astra-db-go/v2/astra"
	"github.com/datastax/astra-db-go/v2/astra/filter"
	"github.com/datastax/astra-db-go/v2/astra/options"
)

func main() {
	ctx := context.Background()

	// Get an existing collection
	client := astra.NewClient()

	database := client.Database(
		"**API_ENDPOINT**",
		options.API().SetToken("**APPLICATION_TOKEN**"),
	)

	collection := database.Collection("**COLLECTION_NAME**")

	// Create the filter
	filterClause := filter.And(
		filter.Eq("is_checked_out", false),
		filter.Lt("number_of_pages", 300),
	)

	// Get the first page
	cursor1 := collection.Find(
		filterClause,
	)

	cursor1.Next(ctx)

	var results1 []map[string]any
	if err := cursor1.DecodeBuffered(&results1, 0); err != nil {
		log.Fatal(err)
	}
	for _, document := range results1 {
		fmt.Println(document)
	}

	if err := cursor1.Err(); err != nil {
		log.Fatal(err)
	}

	paginationState1 := cursor1.NextPageState()

	// Get the next page
	if paginationState1 != nil {
		cursor2 := collection.Find(
			filterClause,
			options.CollectionFind().
				SetInitialPageState(*paginationState1),
		)

		cursor2.Next(ctx)

		var results2 []map[string]any
		if err := cursor2.DecodeBuffered(&results2, 0); err != nil {
			log.Fatal(err)
		}
		for _, document := range results2 {
			fmt.Println(document)
		}

		if err := cursor2.Err(); err != nil {
			log.Fatal(err)
		}

		paginationState2 := cursor2.NextPageState()
		_ = paginationState2 // Can be used for further pagination
	}
}

File: clients/typescript/examples/collections-find-many-iterate-manual.ts
import { DataAPIClient } from "@datastax/astra-db-ts";

// Get an existing collection
const client = new DataAPIClient();
const database = client.db("**API_ENDPOINT**", {
  token: "**APPLICATION_TOKEN**",
});
const collection = database.collection("**COLLECTION_NAME**");

// Create the filter
const filter = {
  $and: [{ is_checked_out: false }, { number_of_pages: { $lt: 300 } }],
};

(async function () {
  // Get the first page
  const cursor1 = collection.find(filter);
  const page1 = await cursor1.fetchNextPage();
  const results1 = page1.result;
  for (const document of results1) {
    console.log(document);
  }
  const paginationState1 = page1.nextPageState;

  // Get the next page
  if (paginationState1) {
    const cursor2 = collection.find(filter, {
      initialPageState: paginationState1,
    });
    const page2 = await cursor2.fetchNextPage();
    const results2 = page2.result;
    for (const document of results2) {
      console.log(document);
    }
    const paginationState2 = page2.nextPageState;
  }
})();
